lib/core/src/can_sendable.rs

```diff
@@ -0,0 +1,292 @@
+        data[3..5].copy_from_slice(&u16_bytes);
+
+        data
+    }
+
+    fn can_decode(can_msg: &[u8; 8]) -> Self {
+        let mut u16_bytes: [u8; 2] = [0; 2];
+        u16_bytes.copy_from_slice(&can_msg[1..3]);
+        let u16_1 = u16::from_le_bytes(u16_bytes);
+
+        u16_bytes.copy_from_slice(&can_msg[3..5]);
+        let u16_2 = u16::from_le_bytes(u16_bytes);
+
+        [u16_1, u16_2]
+    }
+}
+
+impl CanSendable for f32 {
+    fn encode_to_can(&self, board_id: u8) -> [u8; 8] {
+        let mut data: [u8; 8] = [0; 8];
+        data[0] = build_can_header(board_id, CAN_MSG_TYPE_F32);
+
+        let f32_bytes: [u8; 4] = self.to_le_bytes();
+        data[1..5].copy_from_slice(&f32_bytes);
+        data
+    }
+
+    fn can_decode(can_msg: &[u8; 8]) -> Self {
+        let mut f32_bytes: [u8; 4] = [0; 4];
+        f32_bytes.copy_from_slice(&can_msg[1..5]);
+        f32::from_le_bytes(f32_bytes)
+    }
+}
+
+// Struct to hold f32 so they can be sent and decoded over CAN
+pub struct PositionDelta {
+    pub clock: Option<u8>,
+    pub x: Option<f32>,
+    pub y: Option<f32>,
+    pub z: Option<f32>,
+}
+
+impl PositionDelta {
+    // sender side code
+    pub fn new(clock: u8, x: f32, y: f32, z: f32) -> Self {
+        PositionDelta {
+            clock: Some(clock),
+            x: Some(x),
+            y: Some(y),
+            z: Some(z),
+        }
+    }
+
+    // converts a provided F32
+    pub fn encode_to_can(&self, board_id: u8) -> [[u8; 8]; 3] {
+        assert!(
+            self.is_complete(),
+            "Attempted to encode an incomplete PositionDelta"
+        );
+
+        let mut x = self.x.unwrap().encode_to_can(board_id);
+        let mut y = self.y.unwrap().encode_to_can(board_id);
+        let mut z = self.z.unwrap().encode_to_can(board_id);
+
+        x[0] = build_can_header(board_id, CAN_MSG_TYPE_POSDELTA);
+        y[0] = build_can_header(board_id, CAN_MSG_TYPE_POSDELTA);
+        z[0] = build_can_header(board_id, CAN_MSG_TYPE_POSDELTA);
+
+        x[5] = 0;
+        y[5] = 1;
+        z[5] = 2;
+
+        let clock = self.clock.unwrap();
+        x[6] = clock;
+        y[6] = clock;
+        z[6] = clock;
+        [x, y, z]
+    }
+}
+
+impl PositionDelta {
+    // receiver code
+
+    /// Create empty PositionDelta to fill
+    pub fn new_empty() -> Self {
+        PositionDelta {
+            clock: None,
+            x: None,
+            y: None,
+            z: None,
+        }
+    }
+
+    /// Reset position delta to empty
+    pub fn clear(&mut self) {
+        self.clock = None;
+        self.x = None;
+        self.y = None;
+        self.z = None;
+    }
+
+    /// Check we've received a completed PositionDelta, (all x,y,z all with the same clock value)
+    pub fn is_complete(&self) -> bool {
+        self.clock.is_some() && self.x.is_some() && self.y.is_some() && self.z.is_some()
+    }
+
+    /// Returns finished f32
+    pub fn return_complete(&self) -> [f32; 3] {
+        assert!(
```

Comment:
This should return an optional instead so it doesn't crash at runtime, and you then don't need to call is complete, you can just look at whether you get `Some()` back